tokio/src/sync/broadcast.rs

```diff
@@ -0,0 +1,980 @@
+    ///     // At this point, we can abort or continue with lost messages
+    ///
+    ///     assert_eq!(20, rx.recv().await.unwrap());
+    ///     assert_eq!(30, rx.recv().await.unwrap());
+    /// }
+    pub async fn recv(&mut self) -> Result<T, RecvError> {
+        use crate::future::poll_fn;
+
+        poll_fn(|cx| self.poll_recv(cx)).await
+    }
+
+    fn register_waker(&self, cx: &Waker) {
+        self.wait.waker.register_by_ref(cx);
+
+        if !self.wait.queued.load(SeqCst) {
+            // Set `queued` before queuing.
+            self.wait.queued.store(true, SeqCst);
+
+            let mut curr = self.shared.wait_stack.load(SeqCst);
+            let node = Arc::into_raw(self.wait.clone()) as *mut _;
+
+            loop {
+                // Safety: `queued == false` means the caller has exclusive
+                // access to `self.wait.next`.
+                self.wait.next.with_mut(|ptr| unsafe { *ptr = curr });
+
+                let res = self.shared.wait_stack.compare_exchange(curr, node, SeqCst, SeqCst);
+
+                match res {
+                    Ok(_) => return,
+                    Err(actual) => curr = actual,
+                }
+            }
+        }
+    }
+}
+
+impl<T> Drop for Receiver<T> {
+    fn drop(&mut self) {
+        let mut tail = self.shared.tail.lock().unwrap();
+
+        tail.rx_cnt -= 1;
+        let until = tail.pos;
+
+        drop(tail);
+
+        while self.next != until {
+            match self.recv_ref(true) {
+                // Ignore the value
+                Ok(_) => {}
+                // The channel is closed
+                Err(TryRecvError::Closed) => break,
+                // Ignore lagging, we will catch up
+                Err(TryRecvError::Lagged) => {}
+                // Can't be empty
+                Err(TryRecvError::Empty) => panic!("unexpected empty broadcast channel"),
+            }
+        }
+    }
+}
+
+impl<T> Drop for Shared<T> {
+    fn drop(&mut self) {
+        // Clear the wait stack
+        let mut curr = *self.wait_stack.get_mut() as *const WaitNode;
+
+        while !curr.is_null() {
+            let waiter = unsafe { Arc::from_raw(curr) };
+            curr = waiter.next.with(|ptr| unsafe { *ptr });
+        }
+    }
+}
+
+impl<T> fmt::Debug for Sender<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "broadcast::Sender")
+    }
+}
+
+impl<T> fmt::Debug for Receiver<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "broadcast::Receiver")
+    }
+}
+
+impl<T> Slot<T> {
+    /// Try to lock the slot for a receiver. If `false`, then a sender holds the
+    /// lock and the calling task will be notified once the sender has released
+    /// the lock.
+    fn try_rx_lock(&self) -> bool {
+        let mut curr = self.lock.load(SeqCst);
+
+        loop {
+            if curr & 1 == 1 {
+                // Locked by sender
+                return false;
+            }
```

Comment:
TIOLI (i'm not sure if this reads any clearer or not):
```rust
while curr & 1 != 1 {
    // ... loop body ...
} 

false
```

Reply:
I'm personally used to seeing CAS loops structured as:

```rust
loop {
    // conditional break

    // cas

    // check if successful or break.
```